Implement HA scheduling

aaronlehmann opened the pull request:
Change the scheduler to prioritize spreading out tasks from the same
service over spreading out the total number of tasks on each node.

This switches to an O(n) scheduling algorithm, but amortizes the cost of
evaluating all the nodes over a whole batch of tasks that share the same
spec, instead of doing it separately for each task.

ping @dongluochen @aluzzardi

dongluochen commented on manager/scheduler/indexed_node_heap.go:
```diff
@@ -96,6 +99,47 @@ func (nh *nodeHeap) findMin(meetsConstraints func(*NodeInfo) bool, scanAllNodes
+	// do, and cache the result of the constraint evaluation).
+	sort.Sort(sorter{nodes: nodes, lessFunc: nodeLess})
+
+	if len(nodes) < n {
```
Put this in front of `sort.Sort`.

aaronlehmann replied:
I think it's actually necessary to sort the list whether we're returning all the nodes or not. The caller will start with the "best" node at the start of the slice and add tasks to it until that node is no longer "best". The current code isn't doing this yet. I'll fix that.

dongluochen commented on manager/scheduler/scheduler.go:
```diff
@@ -317,26 +317,45 @@ func (s *Scheduler) processPreassignedTasks(ctx context.Context) {
+
+		// Group common tasks with common specs and networks by
+		// marshalling these fields into taskKey and using it as a map
+		// key. Networks are included because PluginFilter makes
```
Can tasks in a service with same spec have different networks?

aaronlehmann replied:
> Can tasks in a service with same spec have different networks?

ping @mrjana - I asked a similar question earlier but I think this is probably a better question to ask.

aaronlehmann commented on manager/scheduler/scheduler.go:
```diff
@@ -427,30 +446,97 @@ func (s *Scheduler) taskFitNode(ctx context.Context, t *api.Task, nodeID string)
+		}
+
+		schedulingDecisions[taskID] = schedulingDecision{old: t, new: &newT}
+		delete(taskGroup, taskID)
```
There is a missing `nodeIter++` here (how on earth do the tests not catch that?)

aaronlehmann replied:
Also fixed. The reason the tests didn't catch it is that they set up a situation where the assignments are supposed to go to particular node, because that node is more lightly loaded.

aaronlehmann commented on manager/scheduler/scheduler.go:
```diff
@@ -427,30 +446,97 @@ func (s *Scheduler) taskFitNode(ctx context.Context, t *api.Task, nodeID string)
-	if err == nil && nodeInfo.addTask(&newT) {
-		s.nodeHeap.updateNode(nodeInfo)
+	nodeIter := 0
+	for taskID, t := range taskGroup {
```
This loop should keep assigning tasks to the current "best" node until it is no longer the best node or doesn't meet the constraints, then move on to the next node (see https://github.com/docker/swarmkit/pull/1446#discussion_r76359670).

aaronlehmann replied:
Fixed.

aaronlehmann commented:
Added a commit that replaces `nodeHeap` with `nodeSet` (a simple map), since the heap property is no longer useful. It also changes `NodeInfo` to keep track of a node's task by service, so it's cheap to check how many tasks from a given service are running on a node.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "ha-sched" git@github.com:aaronlehmann/swarmkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

dperny commented:
Does this replace the old naive spread-scheduling algorithm?

aaronlehmann commented:
> Does this replace the old naive spread-scheduling algorithm?

Yes.

aaronlehmann commented:
Implemented quickselect, added a test. I think this is ready for review.

Adding spec versioning in #1392 will enable a big optimization, but it shouldn't block this PR.

aaronlehmann commented:
@aluzzardi: I've changed this to only consider tasks in desired state `RUNNING` when sorting nodes to choose the best ones for placement. Constraints still consider the resources used by tasks that are in the process of shutting down.

This fixes the issue discussed in #874 for non-resource-constrained scenarios.

dongluochen commented on manager/scheduler/nodeinfo.go:
```diff
@@ -42,15 +51,37 @@ func (nodeInfo *NodeInfo) addTask(t *api.Task) bool {
+			nodeInfo.Tasks[t.ID] = t
+			nodeInfo.DesiredRunningTasksCount++
+			nodeInfo.DesiredRunningTasksCountByService[t.ServiceID]++
+			return true
```
Could `t` and `oldTask` use different amount of resource? Should we update `nodeInfo.AvailableResources`?

aaronlehmann replied:
They must use the same amount of resources, because `t` is an updated version of `oldTask`, and the resource requirements can't change after the task is created.

dongluochen commented on manager/scheduler/nodeinfo.go:
```diff
@@ -42,15 +51,37 @@ func (nodeInfo *NodeInfo) addTask(t *api.Task) bool {
+			nodeInfo.DesiredRunningTasksCountByService[t.ServiceID]--
+			return true
+		}
+		return false
```
from usage of `addTask` function, the return value is not checked. The meaning of the return value is not clear.

aaronlehmann replied:
It is checked in some places. It returns true if `addTask` modified `nodeInfo`, false otherwise. I'll add a comment.

dongluochen commented on manager/scheduler/nodeset.go:
```diff
@@ -0,0 +1,88 @@
+		}
+	}
+
+	if len(nodes) <= n {
```
This could be reduced to:

```
	if len(nodes) > n {
            quickselect.QuickSelect(nodeSorter{nodes: nodes, lessFunc: nodeLess}, n)
            nodes = nodes[:n]
	}

	sort.Sort(nodeSorter{nodes: nodes, lessFunc: nodeLess})
	return nodes
```

dongluochen commented:
LGTM

aluzzardi commented on manager/scheduler/scheduler.go:
```diff
@@ -422,57 +432,118 @@ func (s *Scheduler) taskFitNode(ctx context.Context, t *api.Task, nodeID string)
+		}
+
+		origNodeIter := nodeIter
+		for failedConstraints[nodeIter%len(nodes)] || !s.pipeline.Process(&nodes[nodeIter%len(nodes)]) {
```
Is there an advantage in running the entire pipeline rather than just the resource filter?

aaronlehmann replied:
I suppose the resource filter would be enough, but do you think it's okay for the scheduler to bypass the pipeline and call a specific filter directly?

aluzzardi replied:
Fair enough. I think a few extra cycles won't hurt anyway, the pipeline is pretty cheap so far.

thaJeztah commented:
Did we include this change in the changelog? I was going through the changelogs for 1.13, and couldn't find it (started to doubt if it was included in 1.13 or not) https://github.com/moby/moby/releases/v1.13.0

aaronlehmann commented:
I don't see it in the changelog, but we definitely included this in 1.13.

ronag commented:
This change is a bit under the radar. Is the change of behaviour documented/described anywhere a bit more in depth?
